assets/src/block-editor/helpers/index.js

```diff
@@ -160,6 +173,64 @@ export const addAMPAttributes = ( settings, name ) => {
 	return settings;
 };
 
+/**
+ * Remove AMP deprecated props from any blocks which may contain them.
+ *
+ * @param {Object} settings Block settings.
+ * @param {string} name     Block name.
+ *
+ * @return {Object} Modified block settings.
+ */
+export const removeDeprecatedAmpProps = ( settings, name ) => {
+	const maybeBlockHasDeprecatedProp = Object.keys( mappedAttributes )
+		.some( ( attribute ) => Object.keys( settings.attributes ).includes( attribute ) );
+
+	if ( maybeBlockHasDeprecatedProp ) {
+		// At minimum, the `attributes`, and `save` props are required.
+		const deprecatedConfig = {
+			attributes: settings.attributes,
+			save( blockData ) {
+				const deprecatedProps = {};
+
+				// Render the block.
+				let element = settings.save( blockData );
+				const { attributes: blockAttributes } = blockData;
+
+				// Collect all deprecated props that would be set on the saved HTML block in the post.
+				for ( const [ blockAttributeName, htmlAttributeName ] of Object.entries( mappedAttributes ) ) {
+					if ( blockAttributeName in blockAttributes && blockAttributes[ blockAttributeName ] ) {
+						deprecatedProps[ htmlAttributeName ] = blockAttributes[ blockAttributeName ];
+					}
+				}
+
+				// If there are not any deprecated props, return the unmodified element.
+				if ( 0 === Object.keys( deprecatedProps ).length ) {
+					return element;
+				}
+
+				// Once we know the block will be updated successfully, we can now modify the block's attributes.
+				if ( 'core/gallery' === name ) {
+					settings.attributes.ampCarousel.default = ! select( 'amp/block-editor' ).hasThemeSupport(); // @todo We could just default this to false now even in Reader mode since block styles are loaded.
+				}
```

Comment:
I don't understand _how_ this works. The `default` is changed here, but when is this getting read? It's not part of the below code that does `cloneElement()`. 

This seems like the not ideal place to put this code as it is causing side effects. Something makes me nervous, like is this expected to work properly if old unmigrated posts are pasted into the editor? I just tried it and it seemed to work, but why? 

Humm, and actually, when in legacy Reader mode and I edit a post that originally had:

```html
<!-- wp:gallery {"ids":[1869,770,760],"ampLightbox":true} -->
```

It gets migrated as:

```html
<!-- wp:gallery {"ids":[1869,770,760],"ampCarousel":false,"ampLightbox":true} -->
```

Why is `ampCarousel` false? The call to `select( 'amp/block-editor' ).hasThemeSupport()` is returning `false` so then the default value of the `ampCarousel` attribute should rather be true, right?

Reply:
This isn't taking effect because the `save` callback containing this line runs some time after `settings.attributes` is read, so I think some of this should be abstracted out and run in the filter callback rather than in the `save` function.

Reply:
But it does seem to work. What filter callback could be used?

Reply:
I looked around a bit and am not seeing a good way to do it entirely on the JS side, so I proposed a possible workaround here: https://github.com/ampproject/amp-wp/pull/4775#pullrequestreview-455018758

Reply:
Re https://github.com/ampproject/amp-wp/pull/4775#discussion_r459921643:

> I don't understand _how_ this works. The default is changed here, but when is this getting read? It's not part of the below code that does `cloneElement()`.

Previously, the `ampCarousel` attribute was being defaulted to `true` when using legacy reader mode. So if the `ampCarousel` attribute was not set on the block, that meant `blockAttributes` in the `save()` method will have it set to `true`, and it would then be appended to the `deprecatedProps` object, which should not be the case.

So the reasoning behind the change was to set the `ampCarousel` attribute to always be `false`, and _then_ update the default value we know that the attribute did not have a truthy value set on the block. As I understand, the `save()` method is called when the block is to be parsed and serialized, and the parsing of the block is what we want to filter so it can be updated accordingly.

The change I made in https://github.com/ampproject/amp-wp/pull/4775/commits/6687931e0dbb1c5bc58e6e75202341c594f3e9a5 is indeed hacky and warrants being nervous about; I haven't done much testing with it outside of the bug I was trying to fix. I'll try to see if there is any robust alternative to replicate the fix, but for now I defer to the workaround you provided in https://github.com/ampproject/amp-wp/pull/4775#issuecomment-663390047.

Reply:
I'm pushing up a commit to introduce a new `Service` to formalize that. Let's see how it works…

Reply:
Please check b2901f096. I tried to explain as best I could the reason for the service, trying to recall the history of these attributes.